src/main.rs

```diff
@@ -149,7 +198,27 @@ impl State {
     ///
     /// No change is made to `self` if an error occurs.
     fn do_move(&mut self, mov: Move) -> Result<NextStep, HanoiError> {
-        unimplemented!()
+        let fromPeg = mov.from;
+        let toPeg = mov.to;
+        let topDisk = self.peek_disk(fromPeg);
+        let bottomDisk = self.peek_disk(toPeg);
+
+        if topDisk.is_none() {
+            return Err(HanoiError::EmptyFrom(fromPeg))
+        }
+
+        if bottomDisk.is_some() && topDisk > bottomDisk {
+            return Err(HanoiError::UnstableStack(toPeg, topDisk.unwrap()))
+        }
```

Comment:
It seems you end up checking for unstable stacks here and ignoring the result of `push_disk` (which causes a warning, btw).